### ISSUE TITLE
Add NCCLAllgather feature

### ISSUE THREAD

**Issue by Richie-yan** [horovod/horovod#1943]
Horovod currently supports allgather functions such as mpi allgather and mpi gpu allgather, and lacks nccl allgather. Recently, I'm using horovod to train a visual model. The model will execute allgather once at each step. Now I use MPI_Allgather on CPU. I implemented a simple NCCL_Allgather on GPU in the source code, and the performance has been greatly improved.So this issue plans to provide an ncclallgather implementation.
Two points were found during the implementation process:
1. The ncclAllGather native interface does not support the case where the first dimension is different;
2. For the test of allgather in the test_tensorflow.py script, why does it appear as hvd.allgather on the same gpu device if there is a gpu in the current environment? (For example: test_horovod_allgather function, because if tf.device () is not used artificially, tensorflow will use a GPU for calculation by default)

**Comment by tgaddair** [comment]
Thanks for raising this, @Richie-yan, I think this would be a great contribution.

1. For Horovod, we need an implementation of [allgatherv](https://www.mpich.org/static/docs/latest/www3/MPI_Allgatherv.html).
2. In the Horovod tests for allgather on GPU, we're testing the behavior if the Tensor is on a different GPU for each process (local_rank), but the allgather happens over MPI going through host memory.

**Comment by romerojosh** [comment]
Hi @Richie-yan, thanks for the suggestion and for volunteering to contribute this!

While NCCL doesn't have an allgatherv interface available, the way to accomplish this with NCCL is to do a _grouped_ `NCCL_Broadcast`. This would look something like:
```
ncclGroupStart();
for (int rank = 0; rank < comm_size; ++rank) {
  ncclBroadcast(sendbuf, recvbuf, count, dtype, rank, nccl_comm, stream);
}
ncclGroupEnd();
```
where the `sendbuf`, `recvbuf` and `count` arguments are set to correspond to the data each rank needs to send or receive in the allgatherv call. Feel free to follow up if you have any more questions or need help with the implementation.

**Comment by Richie-yan** [comment]
Hi, @tgaddair @romerojosh 
Compared to ncclAllGather, will ncclBroadcast have any performance loss?
My current approach is to use the ncclAllGather interface. When there are different sizes, it will be downgraded to mpi_allgather. Is this acceptable? Or use ncclBroadcast according to your suggestion?

**Comment by Richie-yan** [comment]
Hi, @tgaddair 
for question 2 mentioned:  In the Horovod tests for allgather on GPU, we're testing the behavior if the Tensor is on a different GPU for each process (local_rank), but the allgather happens over MPI going through host memory.
I still have some doubts: take the test_horovod_allgather function as an example. If the current environment has available GPUs, I use mpirun -np 2 to test the function locally. Although I currently have two ranks, this function will only be executed on one GPU That is, the two rank devices are 0, so the final performance is to do allgather on the tensor of the same device

**Comment by romerojosh** [comment]
@Richie-yan  In the cases where you can use a regular `ncclAllGather` (when the first dimensions are all the same across tensors), it may be faster than using the allgatherv via grouped `ncclBroadcast`. On the other hand, I would expect that the allgatherv using nccl with grouped broadcasts would outperform `MPI_Allgatherv` as it does not require copying the memory to the CPU. 

Since it sounds like you are planning to have a fallback path, I think it'd be better to fallback to the NCCL-based allgatherv implementation rather than to MPI.

**Comment by tgaddair** [comment]
Fixed by #1952.
